[functorch] Refactor life handle storage

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #90317

A "life handle" is a pointer-to-boolean that says whether or not a
TensorWrapper is alive. A TensorWrapper is alive if we are currently
inside of its corresponding transform. An Interpreter is alive if we are
currently inside of its corresponding transform. I.e., for vmap(f)(x),
the BatchedTensor(x, level=1) is alive inside of the execution of f; and
the corresponding VmapInterpreter is alive inside of f.

Previously, there was a global map of level to life handle. It is
possible to get into a state where we have multiple levels that refer to
different Interpreters (if the implementation of an operator calls into
functorch) and that messes up the global map.

This PR changes it so that
- every Interpreter holds a life handle that says if it is alive
- to construct a TensorWrapper, one must either (a) directly pass it a life
handle, or (b) one must create the TensorWrapper when the corresponding
Interpreter is on the stack (and we will automatically grab the life
handle by indexing into the DynamicLayerStack with the level)

(a) is more robust so I changed most of our C++ callsites to do that.
(b) feels a bit hacky to me, but it seems fine for now:
- It'll raise a nice error message if the interpreter isn't on the stack
- all of our Python callsites already follow this convention (we construct
TensorWrappers after pushing the Interpreter onto the stack).

The alternative to (b) is that we always do (a), which we can do in the
future if (b) runs us into any problems.

Test Plan:
- all functorch tests

zou3519 committed:
[functorch] Refactor life handle storage

A "life handle" is a pointer-to-boolean that says whether or not a
TensorWrapper is alive. A TensorWrapper is alive if we are currently
inside of its corresponding transform. An Interpreter is alive if we are
currently inside of its corresponding transform. I.e., for vmap(f)(x),
the BatchedTensor(x, level=1) is alive inside of the execution of f; and
the corresponding VmapInterpreter is alive inside of f.

Previously, there was a global map of level to life handle. It is
possible to get into a state where we have multiple levels that refer to
different Interpreters (if the implementation of an operator calls into
functorch) and that messes up the global map.

This PR changes it so that
- every Interpreter holds a life handle that says if it is alive
- to construct a TensorWrapper, one must either (a) directly pass it a life
handle, or (b) one must create the TensorWrapper when the corresponding
Interpreter is on the stack (and we will automatically grab the life
handle by indexing into the DynamicLayerStack with the level)

(a) is more robust so I changed most of our C++ callsites to do that.
(b) feels a bit hacky to me, but it seems fine for now:
- It'll raise a nice error message if the interpreter isn't on the stack
- all of our Python callsites already follow this convention (we construct
TensorWrappers after pushing the Interpreter onto the stack).

The alternative to (b) is that we always do (a), which we can do in the
future if (b) runs us into any problems.

Test Plan:
- all functorch tests

[ghstack-poisoned]

zou3519 committed:
Update on "[functorch] Refactor life handle storage"


A "life handle" is a pointer-to-boolean that says whether or not a
TensorWrapper is alive. A TensorWrapper is alive if we are currently
inside of its corresponding transform. An Interpreter is alive if we are
currently inside of its corresponding transform. I.e., for vmap(f)(x),
the BatchedTensor(x, level=1) is alive inside of the execution of f; and
the corresponding VmapInterpreter is alive inside of f.

Previously, there was a global map of level to life handle. It is
possible to get into a state where we have multiple levels that refer to
different Interpreters (if the implementation of an operator calls into
functorch) and that messes up the global map.

This PR changes it so that
- every Interpreter holds a life handle that says if it is alive
- to construct a TensorWrapper, one must either (a) directly pass it a life
handle, or (b) one must create the TensorWrapper when the corresponding
Interpreter is on the stack (and we will automatically grab the life
handle by indexing into the DynamicLayerStack with the level)

(a) is more robust so I changed most of our C++ callsites to do that.
(b) feels a bit hacky to me, but it seems fine for now:
- It'll raise a nice error message if the interpreter isn't on the stack
- all of our Python callsites already follow this convention (we construct
TensorWrappers after pushing the Interpreter onto the stack).

The alternative to (b) is that we always do (a), which we can do in the
future if (b) runs us into any problems.

Test Plan:
- all functorch tests

[ghstack-poisoned]

samdow commented on torch/csrc/functorch/init.cpp:
```diff
@@ -449,7 +449,6 @@ void initFuncTorchBindings(PyObject* module) {
   m.def("dump_tensor", &dump_tensor, "dump_tensor");
   m.def("reshape_dim_into", &at::functorch::reshape_dim_into);
   m.def("reshape_dim_outof", &at::functorch::reshape_dim_outof);
-  m.def("are_transforms_active", &at::functorch::areTransformsActive);
```
Through the other refactors did we end up just not using this in the Python? Somehow I remember it being used somewhere but it's probably fine if all the tests are passing

samdow reviewed (approved):
LGTM!

The only thing that makes me a little worried is if the new signature is even strong enough--specifically the onus is on the caller to make sure that current_level and life_handle are from the same interpreter. Should we just pass the interpreter instead? This might also make it easier to change the callsites for `_make_tensor_wrapper`. However, if we want to this all could also be done separately

zou3519 commented:
> Should we just pass the interpreter instead? This might also make it easier to change the callsites for _make_tensor_wrapper. However, if we want to this all could also be done separately

I agree that would be better. I'll file an issue for followup and merge this first to unblock the autograd.Function work.

The reason why I didn't want to handle changing the callsites for `_make_tensor_wrapper` in this PR is that we would need to do some redesign of the functorch python-side API that might rabbithole into changing the python-side API to look like modes:
- How does the developer grab the interpreter? They could grab from the top of the stack (like how pyfunctorch is doing it), or `_grad_increment_nesting` could return an interpreter instead of a level.
- The ideal API for entering/exiting a level is not to use `_grad_increment_nesting/_grad_decrement_nesting`, it is to create a GradInterpreter object (mode context manager?) and then `__enter__` it. To create a GradTensorWrapper, we'd probably just call some method on the GradInterpreter (GradInterpreter.lift) instead of having a function `_wrap_for_grad(Tensor, Interpreter)`.

samdow commented:
> To clarify, are you suggesting that in C++, we change the makeTensorWrapper API to accept an Interpreter instead of a (current_level, life_handle) pair because it's possible that the life handle is from the wrong level?

Ahh yes so there were two parts baked in there. The main suggestion was at the C++ have `makeTensorWrapper` accept an interpreter so that (current_level, life_handle) can never be from different interpreters.

At the Python level, I was thinking that this could mean `_grad_increment_nesting ` could return an interpreter instead of an int (until we get modes). This would mean exposing interpreters to the python API which could be done safely but definitely might be more of a pybind headache than it's worth given that we want to move to modes soon

zou3519 committed:
Update on "[functorch] Refactor life handle storage"


A "life handle" is a pointer-to-boolean that says whether or not a
TensorWrapper is alive. A TensorWrapper is alive if we are currently
inside of its corresponding transform. An Interpreter is alive if we are
currently inside of its corresponding transform. I.e., for vmap(f)(x),
the BatchedTensor(x, level=1) is alive inside of the execution of f; and
the corresponding VmapInterpreter is alive inside of f.

Previously, there was a global map of level to life handle. It is
possible to get into a state where we have multiple levels that refer to
different Interpreters (if the implementation of an operator calls into
functorch) and that messes up the global map.

This PR changes it so that
- every Interpreter holds a life handle that says if it is alive
- to construct a TensorWrapper, one must either (a) directly pass it a life
handle, or (b) one must create the TensorWrapper when the corresponding
Interpreter is on the stack (and we will automatically grab the life
handle by indexing into the DynamicLayerStack with the level)

(a) is more robust so I changed most of our C++ callsites to do that.
(b) feels a bit hacky to me, but it seems fine for now:
- It'll raise a nice error message if the interpreter isn't on the stack
- all of our Python callsites already follow this convention (we construct
TensorWrappers after pushing the Interpreter onto the stack).

The alternative to (b) is that we always do (a), which we can do in the
future if (b) runs us into any problems.

Test Plan:
- all functorch tests

[ghstack-poisoned]

zou3519 commented:
> Ahh yes so there were two parts baked in there. The main suggestion was at the C++ have makeTensorWrapper accept an interpreter so that (current_level, life_handle) can never be from different interpreters.

Cool, yeah, I agree this is good to avoid footguns. I've added that in the latest update.

> At the Python level, I was thinking that this could mean _grad_increment_nesting could return an interpreter instead of an int (until we get modes). This would mean exposing interpreters to the python API which could be done safely but definitely might be more of a pybind headache than it's worth given that we want to move to modes soon

Punting this to the future, I'll file an issue to track it

osalpekar commented:
This seems to be causing all internal contbuilds to fail with the following sign-compare issue:
```
[2022-12-13T16:40:07.590-08:00] Stderr: fbcode/caffe2/aten/src/ATen/functorch/DynamicLayer.cpp:163:32: error: comparison of integers of different signs: 'std::vector::size_type' (aka 'unsigned long') and 'int64_t' (aka 'long') [-Werror,-Wsign-compare]
[CONTEXT] [2022-12-13T16:40:07.590-08:00]       dynamicLayerStack.size() >= level && level >= 1,
[CONTEXT] [2022-12-13T16:40:07.590-08:00]       ~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~
```
More details in [D42019543](https://www.internalfb.com/diff/D42019543).

zou3519 commented:
@osalpekar @dagitses -- can we consider turning on -Wsign-compare in the OSS PyTorch build to avoid things like this in the future?

dagitses commented:
> @osalpekar @dagitses -- can we consider turning on -Wsign-compare in the OSS PyTorch build to avoid things like this in the future?

I would love to do this. There are *a lot* of violations currently though and we'd need a library that can handle sign mismatch arithmetic correctly. I believe there's something proposed for c++20 that we can get a backport of. Basically it will be no cost when it can be (e.g. uint32 compared with int64), but will have a predictable branch when doing things like int32 vs uint64..

https://en.cppreference.com/w/cpp/utility/intcmp

In practice, we can turn this on incrementally throughout the codebase to make it a bit more tractable.

zou3519 commented:
> There are a lot of violations currently though

Is there a reason why the other violations aren't triggering the -Wsign-compare ?
